### ISSUE TITLE
torch.distributions.multinomial.Multinomial cannot be used in batch

### ISSUE THREAD

**Issue by J-zin** [pytorch/pytorch#42407]
torch.distributions.multinomial.Multinomial(total_count=1, probs=None, logits=None, validate_args=None)

the total_count is must be int, but not tensor, so I can't use this function in a batch scenario.  It's quite inconvenient.
![image](https://user-images.githubusercontent.com/30493522/89102992-f7e1fe80-d440-11ea-971b-e111e3cb94c4.png)

**Comment by mruberry** [comment]
Thank you for the feature suggestion, @J-zin! I'll let our distributions experts comment on it.

**Comment by mariosasko** [comment]
Since the `distributions` package follows the design of the `tensorflow_probability` package and the latter supports `total_count` in the form of a number or a list of numbers (the list has to be broadcastable to match the outer dimensions of `probs/logits`, e.g. in Tensorflow `tfd.Multinomial(total_count=[30, 10, 20], probs=[[0.1, 0.9], [0.5, 0.5], [0.2, 0.8]])`, I think it makes sense to add support for tensors.

@mruberry I can make a PR if you agree that this feature is beneficial.

**Comment by mruberry** [comment]
@mariosasko It sounds good to me, @mariosasko! But let's have one of our area experts confirm.

**Comment by fritzo** [comment]
@J-zin How do you intend to use `torch.distributions.Multinomial`?

- If you intend to draw a batch of samples via `.sample()`, then this feature request is blocked by limitations of the underlying [torch.multinomial()](https://pytorch.org/docs/master/generated/torch.multinomial.html) which requires integer total count. Please open a corresponding issue.
- If you intend to score a batch of samples via `.log_prob()`, then you can pass a bogus `total_count` that is only used in validation (and is completely ignored if `validate_args=False`). If you want validation to work you can pass `total_count=total_count.max().item()`. This should be explained in the docs, but please do help us clarify the docs if see a way!

**Comment by firstprayer** [comment]
@fritzo not sure if the fact that torch.multinomial() only takes integer is a blocker for this issue - my understanding is that it would be more convenient to have the total_count be a tensor but under the hook it could still be using torch.multinomial (so there might be some inefficiency but at least the API is more friendly)

**Comment by fritzo** [comment]
@firstprayer thanks, that makes sense. How about we start accepting `LongTensor` `total_count`, and move the `NotImplementedError` from `.__init__()` to `.sample()`?
```diff
diff --git a/torch/distributions/multinomial.py b/torch/distributions/multinomial.py
index 8e43575e10..bb17525331 100644
--- a/torch/distributions/multinomial.py
+++ b/torch/distributions/multinomial.py
@@ -50,8 +50,6 @@ class Multinomial(Distribution):
         return self.total_count * self.probs * (1 - self.probs)

     def __init__(self, total_count=1, probs=None, logits=None, validate_args=None):
-        if not isinstance(total_count, Number):
-            raise NotImplementedError('inhomogeneous total_count is not supported')
         self.total_count = total_count
         self._categorical = Categorical(probs=probs, logits=logits)
         batch_shape = self._categorical.batch_shape
@@ -87,8 +85,15 @@ class Multinomial(Distribution):
         return self._categorical.param_shape

     def sample(self, sample_shape=torch.Size()):
+        total_count = self.total_count
+        if not isinstance(total_count, Number):
+            lb = total_count.min().item()
+            ub = total_count.max().item()
+            if lb != ub:
+                raise NotImplementedError('.sample() does not support inhomogeneous total_count')
+            total_count = lb
         sample_shape = torch.Size(sample_shape)
-        samples = self._categorical.sample(torch.Size((self.total_count,)) + sample_shape)
+        samples = self._categorical.sample(torch.Size((total_count,)) + sample_shape)
         # samples.shape is (total_count, sample_shape, batch_shape), need to change it to
         # (sample_shape, batch_shape, total_count)
         shifted_idx = list(range(samples.dim()))
```

**Comment by richardrl** [comment]
@fritzo How about sampling from a batch of weights? E.g. the "input" to multinomial is num_samples x num_probabilities, to present num_samples different probability distributions

Seems not supported atm, relying on a for loop

**Comment by fritzo** [comment]
@richardrl batching over `probs` or `logits` has always worked. Rather this issue concerns batching over `total_count`.

**Comment by azraelkuan** [comment]
hi, all, any update about batch multinomial for `torch.multinomial`, it can only accept 1d or 2d tensor, which means vector or matrix, but if i have a 3d tensor, like batch x times x dimension, then i can not use this.

**Comment by jykr** [comment]
Has this issue been addressed?

**Comment by fritzo** [comment]
@jykr go for it, we welcome contributions, and I'm happy to review any attempt at a PR 😄

**Comment by dhudsmith** [comment]
@jykr, I'm also interested in this one, and I think I see how to implement it. If you're not working on this, I'd like to go for it.
